docs/pip.md

```diff
@@ -1,264 +1,571 @@
+<!-- TODO add pip example to usage.md? -->
+See also [usage.md](usage.md) for a complete example of Cachi2 usage.
 
-### Per project index
+Cachi2 downloads the Python dependencies into the deps/pip/ subpath of the output directory. The directory is a flat
+list of the downloaded distributions of your runtime and build dependencies.
 
-Your build uses a per project index, specified by the `PIP_INDEX_URL` environment
-variable. This index contains only the source distributions (sdists) for your packages,
-not wheels. That means all packages need to be built from source.
+```text
+cachi2-output/deps/pip
+├── certifi-2022.12.7.tar.gz
+├── ...
+├── pdm-pep517-1.0.6.tar.gz
+├── requests-2.28.2.tar.gz
+├── ...
+└── wheel-0.38.4.tar.gz
+```
+
+To make pip use the downloaded archives, use the [--find-links](https://pip.pypa.io/en/stable/cli/pip_install/#cmdoption-f)
+and [--no-index](https://pip.pypa.io/en/stable/cli/pip_install/#cmdoption-no-index) options. The --find-links option
+tells pip to look for dependency archives in a directory, --no-index prevents pip from preferring PyPI over the local
+directory. Pip also accepts environment variables; Cachi2 generates `PIP_FIND_LINKS` and `PIP_NO_INDEX` for you. See
+[usage: generate environment variables][usage-genenv] for more details.
 
-### Building from source
+### Using external dependencies
 
-When building a package from source, the most important files are typically
-**pyproject.toml** and, if the build backend is setuptools, **setup.cfg** or
-**setup.py**. If you want to know more, check out [PEP 517][pep-517] and
-[PEP 518][pep-518].
+It gets a bit trickier with [external dependencies](#external-dependencies). Pip does not respect the --find-links
+option for dependencies specified via urls. Instead, Cachi2 rewrites your requirements.txt file(s) in place to replace
+the urls with file paths (after you call the `cachi2 inject-files` subcommand). <!-- TODO document in usage doc -->
 
-> :warning: PEPs 518 and 517 were implemented in pip versions 10.0 and 19.0,
-> respectively. Additionally, only pip>=18.0 supports sdists for PEP 518. Older
-> versions of pip will fail to install packages that rely on these PEPs.
+```diff
+- dockerfile-parse @ https://github.com/.../2.0.0.tar.gz \
++ dockerfile-parse @ file:///absolute-path/cachi2-output/deps/pip/.../dockerfile-parse-...tar.gz
+```
 
-If you just want to try installing packages from sdists, you do not need to go through
-Cachito. You can simply pass the `--no-binary :all:` option to pip.
+External dependencies are stored a bit further down the deps/pip tree to avoid mixing them with PyPI dependencies. The
+path and filename is an implementation detail.
+
+```text
+cachi2-output/deps/pip
+├── ...
+├── external-dockerfile-parse
+│   └── dockerfile-parse-external-sha256-36e4469abb0d96b0e3cd656284d5016e8a674cd57b8ebe5af64786fe63b8184d.tar.gz
+└── ...
+```
+
+## Caveats, workarounds
```

Comment:
I think this session ended up being just `workarounds`. :laughing: 

Maybe we can rename it to `Build troubleshooting` or something in this sense, and add a brief intro paragraph like: 
```
This section shows some common errors that can happen when attempting to hermetically build a container using Cachi2's pip prefetched content.
```

Reply:
Renamed to Troubleshooting and added a short intro. It doesn't apply just to builds, for example https://github.com/chmeliik/cachi2/blob/document-pip/docs/pip.md#dependency-does-not-distribute-sources fails during prefetch already